### PR TITLE
dag_run table - replace column conf type - byte ( that store a python pickle ) by a JSON

### LINKED ISSUE #43933: dag_run table - replace column conf type - byte ( that store a python pickle ) by a JSON
### Description

the column `conf` of the table `dag_run` is using the type `byte` ( and storing a python pickle ) on the database , since airflow only support postgres 12+ and mysql 8+ , we could use a `json` type .

And also the user could search "efficiently" in the web-ui a dag_run based on the parameters thanks to that change

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by rawwar**
closes #43933

### CODE REVIEW

File: airflow/models/dagrun.py [hunk 1]
```diff
@@ -140,7 +141,7 @@ class DagRun(Base, LoggingMixin):
     triggered_by = Column(
         Enum(DagRunTriggeredByType, native_enum=False, length=50)
     )  # Airflow component that triggered the run.
-    conf = Column(PickleType)
+    conf = Column(JSONField(json=json), nullable=True)
```
Comment (ephraimbuddy) [inline airflow/models/dagrun.py]: We should probably use JSONB for postgres cc @kaxil

Comment (rawwar) [reply airflow/models/dagrun.py]: I noticed the backfill model was implemented using JSONField. But, I will update

File: airflow/models/dagrun.py [hunk 2]
```diff
@@ -141,7 +142,7 @@ class DagRun(Base, LoggingMixin):
     triggered_by = Column(
         Enum(DagRunTriggeredByType, native_enum=False, length=50)
     )  # Airflow component that triggered the run.
-    conf = Column(PickleType)
+    conf = Column(JSON().with_variant(postgresql.JSONB, "postgresql"))
```
Comment (Lee-W) [inline airflow/models/dagrun.py:145]: will this work for non-postgresql db?

File: airflow/migrations/versions/0050_3_0_0_make_dagrun_conf_as_json.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+            type_=sa.JSON().with_variant(postgresql.JSONB(astext_type=sa.Text()), "postgresql"),
+            existing_nullable=True,
+        )
+
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    """Unapply make dagrun conf as json."""
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (Lee-W) [inline airflow/migrations/versions/0050_3_0_0_make_dagrun_conf_as_json.py]: We'll need to change this after this is marked as ready to review

File: airflow/models/dagrun.py (hunk 2 above)
Comment (rawwar) [reply airflow/models/dagrun.py:145]: Yeah. I checked how this works. Looks like, only for postgres, this setting will be applied. Not for other databases. I also copied this from https://github.com/apache/airflow/pull/44290

**Comment by rawwar** [comment]
@kaxil , Should I be following what you did for XCOM value for DagRun.conf as well? 

You mentioned following in the migration:

```
    # Summary of the change:
    # 1. Create an archived table (`_xcom_archive`) to store the current "pickled" data in the xcom table
    # 2. Extract and archive the pickled data using the condition
    # 3. Delete the pickled data from the xcom table so that we can update the column type
    # 4. Update the XCom.value column type to JSON from LargeBinary/LongBlob
```

**Comment by rawwar** [comment]
So, I'm not sure why CI is all green. But, I tried to verify the alter table query and it isn't actually working. I tried to just select the data in conf data as below

```
select conf, CAST(CONVERT_FROM(conf, 'UTF8') AS JSONB) from dag_run
```

and its throwing 

`error: ERROR:  invalid byte sequence for encoding "UTF8": 0x80`

**Pull request closed without merging**